pkg/acls/s3/storage.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"fmt"
+	"net/url"
+
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/kops/pkg/acls"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/util/pkg/vfs"
+)
+
+// s3PublicAclStrategy is the AclStrategy for objects written to the assets FileRepository
+type s3PublicAclStrategy struct {
+}
+
+var _ acls.ACLStrategy = &s3PublicAclStrategy{}
+
+// GetACL creates a s3PublicAclStrategy object for writing public files with assets FileRepository.
+func (s *s3PublicAclStrategy) GetACL(p vfs.Path, cluster *kops.Cluster) (vfs.ACL, error) {
+	if cluster.Spec.Assets == nil || cluster.Spec.Assets.FileRepository == nil {
+		return nil, nil
+	}
+
+	s3Path, ok := p.(*vfs.S3Path)
+	if !ok {
+		return nil, nil
+	}
+
+	fileRepository := fi.StringValue(cluster.Spec.Assets.FileRepository)
+
+	u, err := url.Parse(fileRepository)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse: %q", fileRepository)
+	}
+
+	// I am not a huge fan of this, but it does work
+	if u.Host != "s3.amazonaws.com" {
+		return nil, nil
+	}
+
+	config, err := url.Parse(cluster.Spec.ConfigStore)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse: %q", fileRepository)
+	}
+
+	if strings.Contains(u.Path, config.Path) {
+		glog.V(8).Infof("path %q is inside of config store %q, not setting public-read acl", u.Path, config.Path)
+		return nil, nil
+	}
+
+	if strings.TrimPrefix(u.Path, "/") == s3Path.Bucket() {
```

Comment:
FileRepository must be at the root of the bucket?  I'm confused as to the expected shape of things here

Reply:
I think some testing will show how this works better.